plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+--    test the shrinking to make sure you don't get in a loop.
+minimalType :: Kind () -> Type TyName DefaultUni ()
+minimalType ty =
+  case ty of
+    Type{} -> unit
+    KindArrow _ k1 k2 ->
+      case k1 : view k2 of
+        [Type{}]         -> list
+        [Type{}, Type{}] -> pair
+        _                -> TyLam () (TyName $ Name "_" (toEnum 0)) k1 $ minimalType k2
+  where
+    view (KindArrow _ k1 k2) = k1 : view k2
+    view _                   = []
+
+    unit = TyBuiltin () (SomeTypeIn DefaultUniUnit)
+    list = TyBuiltin () (SomeTypeIn DefaultUniProtoList)
+    pair = TyBuiltin () (SomeTypeIn DefaultUniProtoPair)
+
+-- | Get the types of builtins at a given kind
+builtinTys :: Kind () -> [SomeTypeIn DefaultUni]
+builtinTys Star =
+  [ SomeTypeIn DefaultUniInteger
+  , SomeTypeIn DefaultUniUnit
+  , SomeTypeIn DefaultUniBool ]
+builtinTys _ = []
+
+-- | Generate "small" types at a given kind such as builtins, bound variables, bound datatypes,
+-- and abstractions /\ t0 ... tn. T
+genAtomicType :: Kind () -> GenTm (Type TyName DefaultUni ())
+genAtomicType k = do
+  tys <- asks geTypes
+  dts <- asks geDatas
+  let atoms = [ TyVar () x | (x, k') <- Map.toList tys, k == k' ] ++
+              [ TyVar () x | (x, Datatype _ (TyVarDecl _ _ k') _ _ _) <- Map.toList dts, k == k' ]
+      builtins = map (TyBuiltin ()) $ builtinTys k
+      lam k1 k2 = do
+        x <- genMaybeFreshTyName "a"
+        TyLam () x k1 <$> bindTyName x k1 (genAtomicType k2)
+  oneofTm $ map pure (atoms ++ builtins) ++ [lam k1 k2 | KindArrow _ k1 k2 <- [k]]
+
+-- | Generate a type at a given kind
+genType :: Kind () -> GenTm (Type TyName DefaultUni ())
+genType k = onSize (min 10) $
+  ifSizeZero (genAtomicType k) $
+    frequencyTm $ [ (1, genAtomicType k) ] ++
+                  [ (2, genFun) | k == Star ] ++
+                  [ (1, genForall) | k == Star ] ++
+                  [ (1, genLam k1 k2) | KindArrow _ k1 k2 <- [k] ] ++
+                  [ (1, genApp) ]
+  where
+    -- this size split keeps us from generating riddiculous types that
+    -- grow huge to the left of an arrow or abstraction (See also the
+    -- genApp case below). This ratio of 1:7 was not scientifically
+    -- established, if you are unhappy about the compleixty of the
+    -- type of arguments that are generated tweaking this might
+    -- be a good idea.
+    genFun = sizeSplit_ 1 7 (genType k) (genType k) (TyFun ())
+
+    genForall = do
+      x <- genMaybeFreshTyName "a"
+      k <- liftGen arbitrary
+      fmap (TyForall () x k) $ onSize (subtract 1) $ bindTyName x k $ genType Star
+
+    genLam k1 k2 = do
+        x <- genMaybeFreshTyName "a"
+        fmap (TyLam () x k1) $ onSize (subtract 1) $ bindTyName x k1 (genType k2)
+
+    genApp = do
+      k' <- liftGen arbitrary
+      sizeSplit_ 1 7 (genType $ KindArrow () k' k) (genType k') $ TyApp ()
+
+-- | Generate a closed type at a given kind
+genClosedType_ :: Kind () -> Gen (Type TyName DefaultUni ())
```

Comment:
I think `_` as a suffix usually means "discards its result", which doesn't seem to be true here.

Reply:
> I think _ as a suffix usually means "discards its result", which doesn't seem to be true here.

I'm not sure what the intention with those underscores was, so I just dropped them, since it didn't cause any name clashes.